BTA-7795 Update collection architecture (#96)
transferzero/api-documentation: _docs/transaction-flow.md

This is the amount that has to be collected from the sender, or funded from the internal balance.

## payin_methods

> When you are processing a collection this object will be used to specify the details of your customer.

Reviewer: ```suggestion
When you are processing a collection this object will be used to specify the details of where you expect the money to come from.
```

### details when collecting money from a customer

{% capture data-raw %}
```javascript
"payin_methods": [
  {
    "type": "GHS::Mobile",
    "in_details": {
      "phone_number": "+2339999999",
      "mobile_provider": "vodafone"
    }
  }
]
```
{% endcapture %}

{% include language-tabbar.html prefix="collection-sender-example" raw=data-raw %}

Please Check the [Collection page]({{ "/docs/collection-details/" | prepend: site.baseurl }}) to see more examples.

### state

The state of the payin method, which can be one of the following:

> * `incomplete`: Some fields need to be filled in in_details before we can initiate the collection request.

Reviewer: ```suggestion
* `incomplete`: Some fields need to be filled in `in_details` before we can initiate the collection request.
```

* `initial`: All required fields in `in_details` are present and collection process with the sender will start.
* `pending`: Collection process has been started, waiting for sender to send funds.
* `success`: Collection succeeded.
* `processing`: Collection succeeded but waiting for final confirmation of receipt.
* `error`: Collection failed. No funds received from sender. You can update or retry the PayinMethod.
* `mispaid`: Collection succeeded but sender sent the wrong amount.
* `canceled`: The transaction has been canceled and we will refund the sender soon.
* `refunded`: The sender has been refunded the amount they sent in.
* `exception`: An exception happened during processing of the collection. Please contact support.

### state_reason_details

> If there is a specific error for the collection, you can find details here.

Reviewer 2: ```suggestion
If there is a specific error for the collection, you can find the details here.
```

> It contains following fields:

Reviewer 2: ```suggestion
It contains the following fields:
```

#### code

Status code of failed collections.

> There are six different categories of errors based on error codes:

Reviewer 2: ```suggestion
There are six different categories of errors based on the following error codes:
```

* 0 - category: `paid` - funds are collected

> * 1x - category: `pending` - Awaiting funds from the sender

Reviewer: I guess this will change once we have the error code investigation. I would defintiely split pending from recipient/sender action required

Author: Yes Zsolt, these have been made as a start

* 3x - category: `temporary_error` - reThe payment provider is not accepting transactions at the moment. We will retry the collection at a later time. You can also edit the payin method or retry the payin

> * 5xx - category: `sender_action_required` - This collection requires an action by the sender

Reviewer: I would definitely not use 5xx codes for sender action required, as that is effectively the mirror of the same thing  as we had recipient action required for payouts, which is 2xx

Author: Very true, I will remove this and we will need to update the base on ProviderUtils as well ([here](https://github.com/bitpesa/provider-utils/blob/f5d542b9b6cba66b0648251956b6cb969b2eff31/lib/provider-utils/payin_method_codes.rb#L33))

> * 3x - category: `temporary_error` - reThe payment provider is not accepting transactions at the moment. We will retry the collection at a later time. You can also edit the payin method or retry the payin

Reviewer 2: ```suggestion
* 3x - category: `temporary_error` - The payment provider is not accepting transactions at the moment. We will retry the collection at a later time. You can also edit the payin method or retry the payin
```
